metals/src/main/scala/scala/meta/internal/parsing/FoldingRangeExtractor.scala

```diff
@@ -217,40 +272,79 @@ final class FoldingRangeExtractor(
             Some(
               range(
                 tree.pos.input,
-                tree.pos.input
-                  .lineToOffset(tree.pos.startLine) - System
-                  .lineSeparator()
-                  .size,
+                parentDefnFoldStart.getOrElse(
+                  tree.pos.input
+                    .lineToOffset(tree.pos.startLine) - System
+                    .lineSeparator()
+                    .size
+                ),
                 tree.pos.end,
               ),
               false, // if braceless add as is
             )
-          } else if (span(block.pos) - 1 > spanThreshold) {
-            Some((tree.pos, true))
           } else {
-            None
+            Some(
+              range(
+                tree.pos.input,
+                parentDefnFoldStart.getOrElse(tree.pos.start),
+                tree.pos.end,
+              ),
+              true,
+            )
           }
         }
         case _: Stat =>
           tree.parent.collect {
-            case _: Term.Block | _: Term.Function | _: Template | _: Defn =>
-              (tree.pos, true)
+            case _: Defn =>
+              parentDefnFoldStart
+                .map(parentFoldStart =>
+                  (
+                    range(
+                      tree.pos.input,
+                      parentFoldStart,
+                      tree.pos.end,
+                    ),
+                    false,
+                  )
+                )
+                .getOrElse((tree.pos, true))
+            case _: Term.Block | _: Term.Function | _: Template =>
+              (tree.pos, false)
           }
         case _ => None
       }
+    }
 
     private def isScala3BlockWithoutOptionalBraces(tree: Tree) = {
-      tree match {
-        case block: Term.Block =>
-          val firstChar = block.pos.input.chars(block.pos.start)
-          block.pos.startLine != block.pos.endLine && firstChar != '{'
-        case temp: Template =>
-          def endChar = temp.pos.input.chars(temp.pos.end)
-          def endsAtEndOfLine = temp.pos.end >= temp.pos.input.chars.size
-          temp.pos.startLine != temp.pos.endLine &&
-          (endsAtEndOfLine || endChar != '}')
-        case _ => false
-      }
+      val childrenTokens = tree.children.flatMap(_.tokens)
+      val blockTokens = tree.parent.get.tokens
+      val LeftBracesNotClosed = blockTokens
+        .slice(0, blockTokens.indexOfSlice(childrenTokens))
+        .filter(t => t.is[Token.LeftBrace] || t.is[Token.RightBrace])
+        .foldLeft(0) {
+          case (n, _: Token.LeftBrace) => n + 1
+          case (n, _: Token.RightBrace) => n - 1
+          case (n, _) => n
+        }
+      LeftBracesNotClosed == 0
+    }
+
+    private def findFoldStartAtParentDefn(tree: Tree, parent: Tree): Int = {
+      val defnTokens = parent.tokens
+      val defnTokenSlice = defnTokens
+        .slice(0, defnTokens.indexOfSlice(tree.tokens))
```

Comment:
```suggestion
        .takeWhile(_.pos.start < tree.pos.start)
```
something along these lines maybe?